# Email Spam Classifier

In [1]:
# Code to convert raw mail to data format

In [2]:
cd 'C:\Users\vinod\Desktop'

In [3]:
# below function will read the content of the data

In [4]:
function [file_contents,ss,rr] = readFile(filename)
%READFILE reads a file and returns its entire contents 
%   file_contents = READFILE(filename) reads a file and returns its entire
%   contents in file_contents
%

% Load File
fid = fopen(filename);
if fid
    [file_contents,ss,rr] = fscanf(fid, '%c', inf);
    fclose(fid);
else
    file_contents = '';
    fprintf('Unable to open %s\n', filename);
end

end


In [5]:
file_contents = readFile('myemailsss2.txt')

file_contents = > Anyone knows how much it costs to host a web portal ?
>
Well, it depends on how many visitors you're expecting.
This can be anywhere from less than 10 bucks a month to a couple of $100. 
You should checkout http://www.rackspace.com/ or perhaps Amazon EC2 
if youre running something big..

To unsubscribe yourself from this mailing list, send an email to:
groupname-unsubscribe@egroups.com




In [ ]:
# I have collected most used email word (1899) and saved in vocab.text
# below function will set index for each word 

In [12]:
function vocabList = getVocabList()
%GETVOCABLIST reads the fixed vocabulary list in vocab.txt and returns a
%cell array of the words
%   vocabList = GETVOCABLIST() reads the fixed vocabulary list in vocab.txt 
%   and returns a cell array of the words in vocabList.


%% Read the fixed vocabulary list
fid = fopen('vocab.txt');

% Store all dictionary words in cell array vocab{}
n = 1899;  % Total number of words in the dictionary

% For ease of implementation, we use a struct to map the strings => integers
% In practice, you'll want to use some form of hashmap
vocabList = cell(n, 1);
for i = 1:n
    % Word Index (can ignore since it will be = i)
    fscanf(fid, '%d', 1);
    % Actual Word
    vocabList{i} = fscanf(fid, '%s', 1);
end
fclose(fid);

end


In [13]:
[vocabList]= getVocabList()

vocabList =
{
  [1,1] = aa
  [2,1] = ab
  [3,1] = abil
  [4,1] = abl
  [5,1] = about
  [6,1] = abov
  [7,1] = absolut
  [8,1] = abus
  [9,1] = ac
  [10,1] = accept
  [11,1] = access
  [12,1] = accord
  [13,1] = account
  [14,1] = achiev
  [15,1] = acquir
  [16,1] = across
  [17,1] = act
  [18,1] = action
  [19,1] = activ
  [20,1] = actual
  [21,1] = ad
  [22,1] = adam
  [23,1] = add
  [24,1] = addit
  [25,1] = address
  [26,1] = administr
  [27,1] = adult
  [28,1] = advanc
  [29,1] = advantag
  [30,1] = advertis
  [31,1] = advic
  [32,1] = advis
  [33,1] = ae
  [34,1] = af
  [35,1] = affect
  [36,1] = affili
  [37,1] = afford
  [38,1] = africa
  [39,1] = after
  [40,1] = ag
  [41,1] = again
  [42,1] = against
  [43,1] = agenc
  [44,1] = agent
  [45,1] = ago
  [46,1] = agre
  [47,1] = agreement
  [48,1] = aid
  [49,1] = air
  [50,1] = al
  [51,1] = alb
  [52,1] = align
  [53,1] = all
  [54,1] = allow
  [55,1] = almost
  [56,1] = alon
  [57,1] = along
  [58,1] = alreadi
  [59,1] = alsa
 

  [455,1] = direct
  [456,1] = directli
  [457,1] = director
  [458,1] = directori
  [459,1] = disabl
  [460,1] = discount
  [461,1] = discov
  [462,1] = discoveri
  [463,1] = discuss
  [464,1] = disk
  [465,1] = displai
  [466,1] = disposit
  [467,1] = distanc
  [468,1] = distribut
  [469,1] = dn
  [470,1] = dnumber
  [471,1] = do
  [472,1] = doc
  [473,1] = document
  [474,1] = doe
  [475,1] = doer
  [476,1] = doesn
  [477,1] = dollar
  [478,1] = dollarac
  [479,1] = dollarnumb
  [480,1] = domain
  [481,1] = don
  [482,1] = done
  [483,1] = dont
  [484,1] = doubl
  [485,1] = doubt
  [486,1] = down
  [487,1] = download
  [488,1] = dr
  [489,1] = draw
  [490,1] = dream
  [491,1] = drive
  [492,1] = driver
  [493,1] = drop
  [494,1] = drug
  [495,1] = due
  [496,1] = dure
  [497,1] = dvd
  [498,1] = dw
  [499,1] = dynam
  [500,1] = ea
  [501,1] = each
  [502,1] = earli
  [503,1] = earlier
  [504,1] = earn
  [505,1] = earth
  [506,1] = easi
  [507,1] = easier
  [508,1] = easili
  [509,1]

  [907,1] = kernel
  [908,1] = kevin
  [909,1] = keyboard
  [910,1] = kid
  [911,1] = kill
  [912,1] = kind
  [913,1] = king
  [914,1] = kingdom
  [915,1] = knew
  [916,1] = know
  [917,1] = knowledg
  [918,1] = known
  [919,1] = la
  [920,1] = lack
  [921,1] = land
  [922,1] = languag
  [923,1] = laptop
  [924,1] = larg
  [925,1] = larger
  [926,1] = largest
  [927,1] = laser
  [928,1] = last
  [929,1] = late
  [930,1] = later
  [931,1] = latest
  [932,1] = launch
  [933,1] = law
  [934,1] = lawrenc
  [935,1] = le
  [936,1] = lead
  [937,1] = leader
  [938,1] = learn
  [939,1] = least
  [940,1] = leav
  [941,1] = left
  [942,1] = legal
  [943,1] = lender
  [944,1] = length
  [945,1] = less
  [946,1] = lesson
  [947,1] = let
  [948,1] = letter
  [949,1] = level
  [950,1] = lib
  [951,1] = librari
  [952,1] = licens
  [953,1] = life
  [954,1] = lifetim
  [955,1] = light
  [956,1] = like
  [957,1] = limit
  [958,1] = line
  [959,1] = link
  [960,1] = linux
  [961,1] = list
  [962,1] = li

  [1333,1] = proven
  [1334,1] = provid
  [1335,1] = proxi
  [1336,1] = pub
  [1337,1] = public
  [1338,1] = publish
  [1339,1] = pudg
  [1340,1] = pull
  [1341,1] = purchas
  [1342,1] = purpos
  [1343,1] = put
  [1344,1] = python
  [1345,1] = qnumber
  [1346,1] = qualifi
  [1347,1] = qualiti
  [1348,1] = quarter
  [1349,1] = question
  [1350,1] = quick
  [1351,1] = quickli
  [1352,1] = quit
  [1353,1] = quot
  [1354,1] = radio
  [1355,1] = ragga
  [1356,1] = rais
  [1357,1] = random
  [1358,1] = rang
  [1359,1] = rate
  [1360,1] = rather
  [1361,1] = ratio
  [1362,1] = razor
  [1363,1] = razornumb
  [1364,1] = re
  [1365,1] = reach
  [1366,1] = read
  [1367,1] = reader
  [1368,1] = readi
  [1369,1] = real
  [1370,1] = realiz
  [1371,1] = realli
  [1372,1] = reason
  [1373,1] = receiv
  [1374,1] = recent
  [1375,1] = recipi
  [1376,1] = recommend
  [1377,1] = record
  [1378,1] = red
  [1379,1] = redhat
  [1380,1] = reduc
  [1381,1] = refer
  [1382,1] = refin
  [1383,1] = reg
  [1384,1]

  [1760,1] = up
  [1761,1] = updat
  [1762,1] = upgrad
  [1763,1] = upon
  [1764,1] = urgent
  [1765,1] = url
  [1766,1] = us
  [1767,1] = usa
  [1768,1] = usag
  [1769,1] = usb
  [1770,1] = usd
  [1771,1] = usdollarnumb
  [1772,1] = useless
  [1773,1] = user
  [1774,1] = usr
  [1775,1] = usual
  [1776,1] = util
  [1777,1] = vacat
  [1778,1] = valid
  [1779,1] = valu
  [1780,1] = valuabl
  [1781,1] = var
  [1782,1] = variabl
  [1783,1] = varieti
  [1784,1] = variou
  [1785,1] = ve
  [1786,1] = vendor
  [1787,1] = ventur
  [1788,1] = veri
  [1789,1] = verifi
  [1790,1] = version
  [1791,1] = via
  [1792,1] = video
  [1793,1] = view
  [1794,1] = virtual
  [1795,1] = visa
  [1796,1] = visit
  [1797,1] = visual
  [1798,1] = vnumber
  [1799,1] = voic
  [1800,1] = vote
  [1801,1] = vs
  [1802,1] = vulner
  [1803,1] = wa
  [1804,1] = wai
  [1805,1] = wait
  [1806,1] = wake
  [1807,1] = walk
  [1808,1] = wall
  [1809,1] = want
  [1810,1] = war
  [1811,1] = warm
  [1812,1] = warn
  [1813,1] = w

In [7]:
# below function will clean the mail content to proper format
# then it will cross check all the words in the email i have give with Vocablist
# if the word in eamil matches the word in vocablist then it will take that index from vocablist

In [14]:
function word_indices = processEmail(email_contents)
%PROCESSEMAIL preprocesses a the body of an email and
%returns a list of word_indices 
%   word_indices = PROCESSEMAIL(email_contents) preprocesses 
%   the body of an email and returns a list of indices of the 
%   words contained in the email. 
%

% Load Vocabulary
vocabList = getVocabList();

% Init return value
word_indices = [];

% ========================== Preprocess Email ===========================

% Find the Headers ( \n\n and remove )
% Uncomment the following lines if you are working with raw emails with the
% full headers

%hdrstart = strfind(email_contents, ([char(10) char(10)]))
%email_contents = email_contents(hdrstart(1):end);

% Lower case
email_contents = lower(email_contents);

% Strip all HTML
% Looks for any expression that starts with < and ends with > and replace
% and does not have any < or > in the tag it with a space
email_contents = regexprep(email_contents, '<[^<>]+>', ' ');

% Handle Numbers
% Look for one or more characters between 0-9
email_contents = regexprep(email_contents, '[0-9]+', 'number');

% Handle URLS
% Look for strings starting with http:// or https://
email_contents = regexprep(email_contents, ...
                           '(http|https)://[^\s]*', 'httpaddr');

% Handle Email Addresses
% Look for strings with @ in the middle
email_contents = regexprep(email_contents, '[^\s]+@[^\s]+', 'emailaddr');

% Handle $ sign
email_contents = regexprep(email_contents, '[$]+', 'dollar');


% ========================== Tokenize Email ===========================

% Output the email to screen as well
fprintf('\n==== Processed Email ====\n\n');

% Process file
l = 0;

while ~isempty(email_contents)

    % Tokenize and also get rid of any punctuation
    [str, email_contents] = ...
       strtok(email_contents, ...
              [' @$/#.-:&*+=[]?!(){},''">_<;%' char(10) char(13)]);
   
    % Remove any non alphanumeric characters
    str = regexprep(str, '[^a-zA-Z0-9]', '');

    % Stem the word 
    % (the porterStemmer sometimes has issues, so we use a try catch block)
    try str = porterStemmer(strtrim(str)); 
    catch str = ''; continue;
    end;

    % Skip the word if it is too short
    if length(str) < 1
       continue;
    end

     

    idx = strmatch(str, vocabList, 'exact');
    if ~isempty(idx)
        word_indices = [word_indices; idx];
    end

 






    % =============================================================


    % Print to screen, ensuring that the output lines are not too long
    if (l + length(str) + 1) > 1000
        fprintf('\n');
        l = 0;
    end
    fprintf('%s ', str);
    l = l + length(str) + 1;



end

% Print footer
fprintf('\n\n=========================\n');

end


In [15]:
word_indices  = processEmail(file_contents)


==== Processed Email ====

anyon know how much it cost to host a web portal well it depend on how mani visitor you re expect thi can be anywher from less than number buck a month to a coupl of dollarnumb you should checkout httpaddr or perhap amazon ecnumb if your run someth big to unsubscrib yourself from thi mail list send an email to emailaddr 

word_indices =

     86
    916
    794
   1077
    883
    370
   1699
    790
   1822
   1831
    883
    431
   1171
    794
   1002
   1893
   1364
    592
   1676
    238
    162
     89
    688
    945
   1663
   1120
   1062
   1699
    375
   1162
    479
   1893
   1510
    799
   1182
   1237
    810
   1895
   1440
   1547
    181
   1699
   1758
   1896
    688
   1676
    992
    961
   1477
     71
    530
   1699
    531



In [18]:
# this is the final step in processing the email
# here X variable is filled with zeros for ...
# ..(1899 row which sholud be equal to vocablist row)
# then in X where ever above word_indicies come it assing 1 in that place

In [16]:
function x = emailFeatures(word_indices)
%EMAILFEATURES takes in a word_indices vector and produces a feature vector
%from the word indices
%   x = EMAILFEATURES(word_indices) takes in a word_indices vector and 
%   produces a feature vector from the word indices. 

% Total number of words in the dictionary
n = 1899;

% You need to return the following variables correctly.
x = zeros(n, 1);

%  

for i=1:size(word_indices)
    x(word_indices(i)) = 1;
end

% =========================================================================
    

end


In [62]:
 features = emailFeatures(word_indices);
 feature_look_like = feature(1:100)
 # i have showed you sample of feature but features variable hold 1899 column of data in that i showed only 100

feature_look_like =

 Columns 1 through 20:

   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0

 Columns 21 through 40:

   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0

 Columns 41 through 60:

   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0

 Columns 61 through 80:

   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0

 Columns 81 through 100:

   0   0   0   0   0   1   0   0   1   0   0   0   0   0   0   0   0   0   0   0



In [20]:
# now our email content is converted 
#and it is ready to apply machine learning (svm )to it

## like this i converted 10000 raw email into above format 
## and saved in two matlab file 
## one is training matlab file which hold 6000 converted email file
## another one test file which hold 4000 converted email file

In [ ]:
# below file contain trainingemail which have 6000 converted email

In [25]:
load('Trainingemail.mat');

In [26]:
# in below C is like regulation parameter it is used in support vector machine
# here i am going to train my data with using (svm  Linear kernel) algorthim

In [30]:
# below function is the svm linear kernel algorithm

In [31]:
function [model] = svmTrain(X, Y, C, kernelFunction, ...
                            tol, max_passes)
%SVMTRAIN Trains an SVM classifier using a simplified version of the SMO 
%algorithm. 
%   [model] = SVMTRAIN(X, Y, C, kernelFunction, tol, max_passes) trains an
%   SVM classifier and returns trained model. X is the matrix of training 
%   examples.  Each row is a training example, and the jth column holds the 
%   jth feature.  Y is a column matrix containing 1 for positive examples 
%   and 0 for negative examples.  C is the standard SVM regularization 
%   parameter.  tol is a tolerance value used for determining equality of 
%   floating point numbers. max_passes controls the number of iterations
%   over the dataset (without changes to alpha) before the algorithm quits.
%
% Note: This is a simplified version of the SMO algorithm for training
%       SVMs. In practice, if you want to train an SVM classifier, we
%       recommend using an optimized package such as:  
%
%           LIBSVM   (http://www.csie.ntu.edu.tw/~cjlin/libsvm/)
%           SVMLight (http://svmlight.joachims.org/)
%
%

if ~exist('tol', 'var') || isempty(tol)
    tol = 1e-3;
end

if ~exist('max_passes', 'var') || isempty(max_passes)
    max_passes = 5;
end

% Data parameters
m = size(X, 1);
n = size(X, 2);

% Map 0 to -1
Y(Y==0) = -1;

% Variables
alphas = zeros(m, 1);
b = 0;
E = zeros(m, 1);
passes = 0;
eta = 0;
L = 0;
H = 0;

% Pre-compute the Kernel Matrix since our dataset is small
% (in practice, optimized SVM packages that handle large datasets
%  gracefully will _not_ do this)
% 
% We have implemented optimized vectorized version of the Kernels here so
% that the svm training will run faster.
if strcmp(func2str(kernelFunction), 'linearKernel')
    % Vectorized computation for the Linear Kernel
    % This is equivalent to computing the kernel on every pair of examples
    K = X*X';
elseif strfind(func2str(kernelFunction), 'gaussianKernel')
    % Vectorized RBF Kernel
    % This is equivalent to computing the kernel on every pair of examples
    X2 = sum(X.^2, 2);
    K = bsxfun(@plus, X2, bsxfun(@plus, X2', - 2 * (X * X')));
    K = kernelFunction(1, 0) .^ K;
else
    % Pre-compute the Kernel Matrix
    % The following can be slow due to the lack of vectorization
    K = zeros(m);
    for i = 1:m
        for j = i:m
             K(i,j) = kernelFunction(X(i,:)', X(j,:)');
             K(j,i) = K(i,j); %the matrix is symmetric
        end
    end
end

% Train
fprintf('\nTraining ...');
dots = 12;
while passes < max_passes,
            
    num_changed_alphas = 0;
    for i = 1:m,
        
        % Calculate Ei = f(x(i)) - y(i) using (2). 
        % E(i) = b + sum (X(i, :) * (repmat(alphas.*Y,1,n).*X)') - Y(i);
        E(i) = b + sum (alphas.*Y.*K(:,i)) - Y(i);
        
        if ((Y(i)*E(i) < -tol && alphas(i) < C) || (Y(i)*E(i) > tol && alphas(i) > 0)),
            
            % In practice, there are many heuristics one can use to select
            % the i and j. In this simplified code, we select them randomly.
            j = ceil(m * rand());
            while j == i,  % Make sure i \neq j
                j = ceil(m * rand());
            end

            % Calculate Ej = f(x(j)) - y(j) using (2).
            E(j) = b + sum (alphas.*Y.*K(:,j)) - Y(j);

            % Save old alphas
            alpha_i_old = alphas(i);
            alpha_j_old = alphas(j);
            
            % Compute L and H by (10) or (11). 
            if (Y(i) == Y(j)),
                L = max(0, alphas(j) + alphas(i) - C);
                H = min(C, alphas(j) + alphas(i));
            else
                L = max(0, alphas(j) - alphas(i));
                H = min(C, C + alphas(j) - alphas(i));
            end
           
            if (L == H),
                % continue to next i. 
                continue;
            end

            % Compute eta by (14).
            eta = 2 * K(i,j) - K(i,i) - K(j,j);
            if (eta >= 0),
                % continue to next i. 
                continue;
            end
            
            % Compute and clip new value for alpha j using (12) and (15).
            alphas(j) = alphas(j) - (Y(j) * (E(i) - E(j))) / eta;
            
            % Clip
            alphas(j) = min (H, alphas(j));
            alphas(j) = max (L, alphas(j));
            
            % Check if change in alpha is significant
            if (abs(alphas(j) - alpha_j_old) < tol),
                % continue to next i. 
                % replace anyway
                alphas(j) = alpha_j_old;
                continue;
            end
            
            % Determine value for alpha i using (16). 
            alphas(i) = alphas(i) + Y(i)*Y(j)*(alpha_j_old - alphas(j));
            
            % Compute b1 and b2 using (17) and (18) respectively. 
            b1 = b - E(i) ...
                 - Y(i) * (alphas(i) - alpha_i_old) *  K(i,j)' ...
                 - Y(j) * (alphas(j) - alpha_j_old) *  K(i,j)';
            b2 = b - E(j) ...
                 - Y(i) * (alphas(i) - alpha_i_old) *  K(i,j)' ...
                 - Y(j) * (alphas(j) - alpha_j_old) *  K(j,j)';

            % Compute b by (19). 
            if (0 < alphas(i) && alphas(i) < C),
                b = b1;
            elseif (0 < alphas(j) && alphas(j) < C),
                b = b2;
            else
                b = (b1+b2)/2;
            end

            num_changed_alphas = num_changed_alphas + 1;

        end
        
    end
    
    if (num_changed_alphas == 0),
        passes = passes + 1;
    else
        passes = 0;
    end

    fprintf('.');
    dots = dots + 1;
    if dots > 78
        dots = 0;
        fprintf('\n');
    end
    if exist('OCTAVE_VERSION')
        fflush(stdout);
    end
end
fprintf(' Done! \n\n');

% Save the model
idx = alphas > 0;
model.X= X(idx,:);
model.y= Y(idx);
model.kernelFunction = kernelFunction;
model.b= b;
model.alphas= alphas(idx);
model.w = ((alphas.*Y)'*X)';

end


In [32]:
function sim = linearKernel(x1, x2)
%LINEARKERNEL returns a linear kernel between x1 and x2
%   sim = linearKernel(x1, x2) returns a linear kernel between x1 and x2
%   and returns the value in sim

% Ensure that x1 and x2 are column vectors
x1 = x1(:); x2 = x2(:);

% Compute the kernel
sim = x1' * x2;  % dot product

end

In [33]:
C = 0.1;
model = svmTrain(X, y, C, @linearKernel);


Training ......................................................................
...............................................................................
.................................................................. Done! 



# now we need to test our model with same training data to see how good our model classifying the mail
# in our traing dataset we already classified the mail which is hold in Y column of training dataset (as 0 and 1) 0= not spam 1=spam
# now our model will give predicted(Y) for given training email (as 0 and 1) 0= not sapm 1=spam
# now we should cross check Y with predicted Y


In [ ]:
# below function make use of model to get predicted Y

In [34]:
function pred = svmPredict(model, X)
%SVMPREDICT returns a vector of predictions using a trained SVM model
%(svmTrain). 
%   pred = SVMPREDICT(model, X) returns a vector of predictions using a 
%   trained SVM model (svmTrain). X is a mxn matrix where there each 
%   example is a row. model is a svm model returned from svmTrain.
%   predictions pred is a m x 1 column of predictions of {0, 1} values.
%

% Check if we are getting a column vector, if so, then assume that we only
% need to do prediction for a single example
if (size(X, 2) == 1)
    % Examples should be in rows
    X = X';
end

% Dataset 
m = size(X, 1);
p = zeros(m, 1);
pred = zeros(m, 1);

if strcmp(func2str(model.kernelFunction), 'linearKernel')
    % We can use the weights and bias directly if working with the 
    % linear kernel
    p = X * model.w + model.b;
elseif strfind(func2str(model.kernelFunction), 'gaussianKernel')
    % Vectorized RBF Kernel
    % This is equivalent to computing the kernel on every pair of examples
    X1 = sum(X.^2, 2);
    X2 = sum(model.X.^2, 2)';
    K = bsxfun(@plus, X1, bsxfun(@plus, X2, - 2 * X * model.X'));
    K = model.kernelFunction(1, 0) .^ K;
    K = bsxfun(@times, model.y', K);
    K = bsxfun(@times, model.alphas', K);
    p = sum(K, 2);
else
    % Other Non-linear kernel
    for i = 1:m
        prediction = 0;
        for j = 1:size(model.X, 1)
            prediction = prediction + ...
                model.alphas(j) * model.y(j) * ...
                model.kernelFunction(X(i,:)', model.X(j,:)');
        end
        p(i) = prediction + model.b;
    end
end

% Convert predictions into 0 / 1
pred(p >= 0) =  1;
pred(p <  0) =  0;

end



In [ ]:
# below is the Predicted y (0= not spam 1= spam)

In [65]:
p = svmPredict(model, X);
p_look_like=p(1:20)

p_look_like =

   1
   1
   0
   0
   0
   0
   1
   0
   0
   0
   1
   0
   0
   0
   1
   0
   0
   0
   1
   1



In [ ]:
# below is y from training dataset (0=not spam 1=spam)

In [66]:
y;
y_look_like=y(1:20)

y_look_like =

   1
   1
   0
   0
   0
   0
   1
   0
   0
   0
   1
   0
   0
   0
   1
   0
   0
   0
   1
   1



In [37]:
# now need to cross check predicted y with y from training dataset ...
# .. to find accuracy of our model working

In [38]:
accuracy = mean(double(p == y)) * 100

accuracy =  99.800


# Now need to check model with test dataset

In [39]:
load('Testemail.mat');

In [40]:
# here we use testemali data to see how good our model classfiy our test email

In [68]:
p = svmPredict(model, Xtest);
p_look_like=p(1:20)

p_look_like =

   1
   0
   0
   1
   1
   1
   1
   0
   1
   0
   0
   0
   0
   0
   1
   0
   0
   1
   1
   1



In [42]:
accuracy_of_model_in_testdata=mean(double(p == ytest)) * 100

accuracy_of_model_in_testdata =  98.800


#  Now model is ready to classify mail
#  now we can interface this model into app .Then that app is capable to classify your mail with 98.80 accuracy 

In [43]:
# testing this model with 2 mail from my inbox let see how it works

In [44]:
file_contents = readFile('myemailsss.txt')

file_contents = rom:	View Details Now <newsletters@shopatbest.com>
reply-to:	View Details Now <newsletters@shopatbest.com>
to:	saeccapamerica05079@gmail.com
date:	29 Jan 2019, 17:04
subject:	Welcome To Unconventional Resorts Perks of Joining ClubM
mailed-by:	emm.freyaexports.com
Signed by:	shopatbest.com
unsubscribe:	Unsubscribe from this sender
security:	Standard encryption (TLS) Learn more

Free* 6N/7D Singapore holiday, includes 3N Cruise trip with membership.
Various other benefits of becoming a Club Mahindra Member:
    7 days of holidays every year for 25 years
    50+ resorts in India & abroad to choose from
    Various in-resort activities  adventure, nature trails, cultural entertainment & more
Join the Club Now!
You can simply WIPE Out your email from Mailing List Tap Here


In [48]:
word_indices  = processEmail(file_contents);


==== Processed Email ====

rom view detail now repli to view detail now to emailaddr date number jan number number number subject welcom to unconvent resort perk of join clubm mail by emm freyaexport com sign by shopatbest com unsubscrib unsubscrib from thi sender secur standard encrypt tl learn more free numbern numberd singapor holidai includ numbern cruis trip with membership variou other benefit of becom a club mahindra member number dai of holidai everi year for number year number resort in india abroad to choos from variou in resort activ adventur natur trail cultur entertain more join the club now you can simpli wipe out your email from mail list tap here 



In [49]:
x = emailFeatures(word_indices);

In [47]:
p = svmPredict(model, x)

p =  1


# above mail is spam (can anyone offer free trip to singapore) my model predicted correctly

In [50]:
# now we check with another mail

In [51]:
file_contents = readFile('myemailsss2.txt')

file_contents = from:	Dan Bader <mail@dbader.org>
to:	saeccapamerica05079@gmail.com
date:	1 Feb 2019, 11:00
subject:	5 tweaks for your Python dev setup you should make today
mailed-by:	drip.dbader.org
Signed by:	dbader.org
security:	Standard encryption (TLS) Learn more
 Hey there,

Ive been using the Sublime Text editor for almost 4 years now in my Python development workflow

And I think its an *amazing* combo.

There are many other editors and IDEs available for writing Python (and I don't "discriminate" against any of them.)

But Sublime just hits the sweet spot for me after having used PyCharm, Emacs, Atom, TextMate and a bunch of other editors.

I kept getting questions about this development setup when I used it in my YouTube screencasts.

And it made me realize how difficult it can be to set up an enjoyable Python development environmentone that's fun to work with.

There are 5 simple tweaks you can apply to make your own Python development environment better: more efficient and

In [52]:
word_indices  = processEmail(file_contents);


==== Processed Email ====

from dan bader to emailaddr date number feb number number number subject number tweak for your python dev setup you should make todai mail by drip dbader org sign by dbader org secur standard encrypt tl learn more hei there iv been us the sublim text editor for almost number year now in my python develop workflow and i think it an amaz combo there ar mani other editor and id avail for write python and i don t discrimin against ani of them but sublim just hit the sweet spot for me after have us pycharm emac atom textmat and a bunch of other editor i kept get question about thi develop setup when i us it in my youtub screencast and it made me realiz how difficult it can be to set up an enjoy python develop environmenton that s fun to work with there ar number simpl tweak you can appli to make your own python develop environ better more effici and more enjoy to us my free number sublim text tweak to boost your python product email cours show you how if you re i

In [53]:
x = emailFeatures(word_indices);

In [54]:
p = svmPredict(model, x)

p = 0


# Above mail is from my python instructor and it is not spam my model predicted it correctly

In [55]:
# you can also improve the classifier with image containing email

In [ ]:
if any doubt reach me with my github